### PR TITLE
Create adjustable test settings.

### PR DESCRIPTION
**Pull request by polina-c**
Contributes to https://github.com/flutter/flutter/issues/135856

Draft of PR for Flutter to consume the changes: https://github.com/flutter/flutter/pull/136692.

After merging this PR, I will update flutter/test, to have 8% of tests that are not leak free yet looking like this: 

```
    testWidgets('initialTimerDuration falls within limit',
    experimentalLeakTracking: LeakTrackingForTests.ignored(),
    (WidgetTester tester) async {
        ...
```

### CODE REVIEW

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:108]: how about `.track([...notGCed, ...classes])` so we remove a call to `track`? same below.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:101]: naming: `copyWithTrackedLeaks` to be consistent with name suggestion above

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:120]: r/track/`trackLeaks` to be consistent with opposite `allowLeaks`

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:136]: should this be private?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:172]: wording nit: Callback to invoke before the test fails when [failOnLeaks] is true and leaks were found.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:173]: nit: onLeaksFound? onLeaksCollected?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:176]: allowedLeaks

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:178]: sp: diagnostic

**Review by kenzieschmoll — COMMENTED** [review]
Some comments around naming that might make these classes / method names more intuitive.

File: pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart
Comment (polina-c) [reply pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart:44]: In general I agree allowed is a standard.

In this concrete case when I worked with leaks, I noticed the word 'allow', that is used now to exempt leaks in FF, is confusing: do we want to allow leaks to happen or allow exception about the leak to happen.

The word 'skip' is preciser in this situation, as it means we do not want to notice the leak.

How does it sound?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:136]: It cannot be instantiated by user, but user can see the instance and create other instances out of it.

File: pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart
Comment (polina-c) [reply pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart:44]: The same relates to `remove` and `contains`. 
Remove from leaks we want to track or remove from allow/skip list?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart]: yes, thanks, added test coverage

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart]: added `copy` and `Leaks`

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart]: added `Leaks`

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:24]: I did not use name `withPaused` and `withStarted` because this method will be used massively, to exempt individual tests from leak tracking.

After merging this PR, I will update flutter/test, to have 8% of tests that are not leak free yet looking like this: 

```
    testWidgets('initialTimerDuration falls within limit',
    experimentalLeakTracking: LeakTrackingForTests.paused(),
    (WidgetTester tester) async {
        ...
```

withPaused looks less readable.

File: pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart:111]: nit: update dart-doc. "The total set of skipped leaks for both [nonGCed] and [notDisposed] leaks."

File: pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart:44]: nit: SkippedLeakSet to be more consistent with `SkippedLeaks`?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:24]: wdyt about replacing "paused" with "ignore" or "skip" and replacing "started" with "track"? Pause and start are confusing to me because this implies some sort of live and modifiable state, whereas it is actually a static setting defined by code. This would also be consistent with the "skip" and "track" terminology used elsewhere which is nice.

File: pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart
Comment (kenzieschmoll) [reply pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart]: I agree this isn't a simple feature set to name... reviewed your changes and what you have looks good. Just a few more nits.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (goderbauer) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:1]: Should this file be exported by `lib/leak_tracker_flutter_testing.dart` or is it just for package-internal use?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (goderbauer) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:24]: +1 to this

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (goderbauer) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:11]: It is not entirely clear to me how this class is supposed to be used (some docs (incl. usage examples)) would help).

in the example given in the comment it appears to be used to configure a `testWidgets` test case - I am confused that it contains mutable state, though. Can you change the test configuration while a test is running? Why is that necessary?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (goderbauer) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:13]: Why is this a global state singleton? Shouldn't I be able to configure this for each test separately? And should this field be user-writable instead of requireing users to go through the defined methods below?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (goderbauer) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:23]: This API is a bit confusing. It is not clear to me as a user what the difference between calling `pause` and `paused` is supposed to be.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (goderbauer) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:11]: Why does this class with only static members exist in the first place? https://dart.dev/effective-dart/design#avoid-defining-a-class-that-contains-only-static-members

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:24]: good idea

File: pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart
Comment (polina-c) [reply pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart:45]: Added comment.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:33]: Added comment for the class.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:37]: Comment for the class may help.
I do not want users to create the settings from scratch, because I want defaults for package to affect all the test. So that if you set leak tracking to paused for the entire package, it would stop all the tests, except those that explicitly set paused=false, in spite of default.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart]: Comment for the class should help.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:33]: The rule states 'AVOID', not 'DON'T'.

This rule is violated across Dart and Flutter many times (https://api.dart.dev/stable/3.0.6/dart-developer/Service/Service.html).

I think we need to convince language team to provide a way to group static methods, so that we do not use abstract classes for these purposes.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart]: And `paused` is renamed to `ignore` now.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:44]: these need to be renamed to ignore and track

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:48]: should these usages of paused be changed to ignore?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (kenzieschmoll) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:52]: same comment here

**Review by kenzieschmoll — APPROVED** [review]
lgtm from me after cleaning up lingering "pause" names.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:44]: There are two static methods: 
ignore() - returns instance of settings with paused set to true
pause - updates current settings with paused set to true

It looks meaningful for me, because first one is intended to skip just one test, while second one stops leak tracking for any test, that uses default.

We can think about other names, but we need two different names to avoid collision.

Comment (kenzieschmoll) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:44]: Based on the comment thread here, https://github.com/dart-lang/leak_tracker/pull/152#discussion_r1361003657, I thought we agreed to replace `pause` and `start` with `ignore` and `track`? With all four terminologies in the mix this may be difficult for a user to understand the API. It is confusing to have 4 terms instead of just 2 that are consistent throughout.

Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:44]: Okay, I see, you suggest to replace it everywhere:

SkippedLeaks (and all its members) -> IgnoredLeaks 
LeakTrackingForTestsSettings.pause -> LeakTrackingForTestsSettings.ignore
Static methods `ignore` (update settings) and `ignored` (return copy of settings with ignore = true)

Comment (kenzieschmoll) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:44]: > SkippedLeaks (and all its members) -> IgnoredLeaks

This wasn't part of my original suggestion, but it could make sense to keep terminology consistent throughout this API. @goderbauer do you have an opinion on naming here?

My comment was regarding current uses of `pause` and `start`. So
`LeakTrackingForTestsSettings.pause -> LeakTrackingForTestsSettings.ignore` and 
`LeakTrackingForTestsSettings.start -> LeakTrackingForTestsSettings.track`.

Also changing the `copyWith` parameter name `paused` to `ignored` or maybe `ignoreAll` would be better. Same for the class member `paused`.

Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart]: Updated everything to 'ignore'. @goderbauer , I am still open for other suggestions. Naming is important and it is better to make it right now.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (goderbauer) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:29]: The API and the way this class can be used feel super complex. Are all the different ways in which this class can be used necessary? What are the use cases that this enables?

Naively, I would have expected that we need to support two use cases:
1) Setting a default setting for all `testWidgets` that don't configure a local setting.
2) On a per test basis, `testsWidgets` should be able to override the defaults.

For that I would have expected something simple like: I can instantiate an immutable `LeakTrackingSettings` object. It's either configured by setting arguments in the `LeakTrackingSettings` constructor or I can invoke named constructors for some well known use cases (e.g. to track all leaks). I can then either pass that constructed object to `testWidgets` to configure something on a test-by-test basis, or I assign that object to a static (e.g. `LeakTrackingSettings.defaults`) to configure the defaults for all tests (e.g. from `flutter_test_config.dart`). 

The immutable `LeakTrackingSettings` object could have a copyWith method to create a setting derived from another. For example, if I just want to change one thing for a given test from the default settings I would do something like:

```dart
testWidgets('foo', leakTracking: LeakTrackingSettings.defaults.copyWith(barSetting: false), ...);
```

There could even be convenience copy methods for well-known use cases.

This would give `LeakTrackingSettings` a single, easy-to-understand API and it would make it explicit from what setting something is derived from.

Note: I am using `LeakTrackingSettings` as a placeholder name in the above. It could be called something else.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (goderbauer) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:32]: This doesn't seem to match the implementation since settings is just a field?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (goderbauer) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:173]: Why is that a separate class then instead of just defining these on LeakTrackingForTestsSettings directly?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (goderbauer) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart]: It feels like LeakTrackingForTestsSettings and LeakTrackingForTests should just be one class then? Why do the statics need to be separated out?

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:32]: For now nothing matches implementation, because leak tracker does not use the settings yet.
It is work in progress.

File: pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart
Comment (polina-c) [reply pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart:111]: Added comment.

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart
Comment (polina-c) [reply pkgs/leak_tracker_flutter_testing/lib/src/leak_tracking_for_tests.dart:32]: Updated comment.

File: pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart
Comment (goderbauer) [reply pkgs/leak_tracker/lib/src/leak_tracking/_primitives/model.dart:114]: (Not sure what use this API has if it doesn't consider the limit 🤷)

File: pkgs/leak_tracker_flutter_testing/lib/src/leak_testing.dart
Comment (goderbauer) [inline pkgs/leak_tracker_flutter_testing/lib/src/leak_testing.dart:26]:
```suggestion
///     leakTracking: LeakTesting.settings.withIgnoredAll(),
```

File: pkgs/leak_tracker_flutter_testing/test/tests/leak_tracking_for_tests_test.dart
Comment (goderbauer) [inline pkgs/leak_tracker_flutter_testing/test/tests/leak_tracking_for_tests_test.dart:63]: Could you do the following and then remove `@visibleForTesting` from the LeakTesting constructor and make it actually private to underline that nobody should be instantiating it directly?

```suggestion
  late LeakTesting defaults;

  setUp(() {
    defaults = LeakTesting.settings;
  });
  tearDown(() {
    LeakTesting.settings = defaults;
  });
```

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**
